dask/bag/core.py

```diff
@@ -756,6 +764,98 @@ def write(data, filename):
         f.close()
 
 
+def _get_s3_bucket(bucket_name, aws_access_key, aws_secret_key, connection,
+                   anon):
+    """Connect to s3 and return a bucket"""
+    import boto
+    if anon is True:
+        connection = boto.connect_s3(anon=anon)
+    elif connection is None:
+        connection = boto.connect_s3(aws_access_key, aws_secret_key)
+    return connection.get_bucket(bucket_name)
+
+
+# we need an unmemoized function to call in the main thread. And memoized
+# functions for the dask.
+_memoized_get_bucket = toolz.memoize(_get_s3_bucket)
+
+
+def _get_key(bucket_name, conn_args, key_name):
+    bucket = _memoized_get_bucket(bucket_name, *conn_args)
+    key = bucket.get_key(key_name)
+    ext = key_name.split('.')[-1]
+    return stream_decompress(ext, key.read())
+
+
+def _parse_s3_URI(bucket_name, paths):
+    # FIXME
+    # DIRTY HACK: so that urlparse won't think '?' in the URI path is for a
+    # query. We replace '?' with 'QUESTIONMARK' so it stays in the path
+    # then after urlpars gives us the path replace 'QUESTIONMARK' with '?'.
+    # So this can be made to fail if a path has 'QUESTIONMARK' in it.
+    if '?' in bucket_name:
+        bucket_name = bucket_name.replace('?', 'QUESTIONMARK')
+    o = urlparse(bucket_name)
+    # if path is specified
+    if (paths == '*') and (o.path != '' and o.path != '/'):
+        paths = o.path[1:].replace('QUESTIONMARK', '?')
+    bucket_name = o.hostname
+    return bucket_name, paths
```

Comment:
I have a less hacky way of doing this. @cowlicks Should I submit a PR to your branch?